### ISSUE TITLE
Auto-save unsaved files

### ISSUE THREAD

**Issue by weeblr** [Microsoft/vscode#906]
A very handy feature in Notepad++: one can just open a new editing tab, or multiple ones, cut and paste stuff, or scribble a few notes. 
Upon closing the application or shutting down the computer, instead of the usual "Do you want to save/not save/cancel" (as vsCode does), all those unsaved and unnamed files are simply saved to a temp location by the app, and restored in the same state upon next start.

Rgds

**Comment by bluecaret** [comment]
I would like to second this request. If this is to compete at all with Sublime and others, this needs to be added. Makes it so much more useful and I feel "safer" using it if I don't have to worry about loosing files.

**Comment by pascon** [comment]
In the [docs ](https://code.visualstudio.com/docs/editor/codebasics#_saveauto-save)there is information about `files.autoSave: "onFocusChange"`:

> onFocusChange to save files when focus moves out of the editor of the dirty file.

Is this option working?

**Comment by bpasero** [comment]
Yes

**Comment by pascon** [comment]
@bpasero that strange, maybe I'm misunderstanding this option but when I try following steps:

1. Open new file
2. Write some text
3. Try closing vscode

file is not saved as dirty and vscode asks me for saving the file. 

My settings.json file looks like below:
`{
    "files.autoSave": "onFocusChange"
}`

Isn't this flag enables auto-saving as in Notepad++ described in first post in this issue?

**Comment by bpasero** [comment]
@pascon works for me, please attach a video of what you try and see.

**Comment by alexandrudima** [comment]
@bpasero I think this is about untitled files

**Comment by bpasero** [comment]
Ok, for untitled files we need to ask the user which filename to pick.

**Comment by weeblr** [comment]
HI

@bpasero 

> Ok, for untitled files we need to ask the user which filename to pick.

No, that's the point. Don't ask anything. Make up a temp file name, store and reopen when vscode is started next time.

See my OP:

> instead of the usual "Do you want to save/not save/cancel" (as vsCode does), all those unsaved and unnamed files are simply saved to a temp location by the app, and restored in the same state upon next start.

Cheers

**Comment by bpasero** [comment]
@weeblr why do you assume this feature works if the feature request is still open?

**Comment by weeblr** [comment]
@bpasero Sorry, I'm confused. From your previous comment, I understood you wanted to implement that feature by prompting the user for a name, in case the file didn't have one already. So I replied that that's not how it should be implemented, per my request.
Isn't it what you suggested?

Rgds

**Comment by bpasero** [comment]
@weeblr sorry I confused you with @pascon

**Comment by weeblr** [comment]
@bpasero ok, no problem, understand now.

**Comment by pascon** [comment]
@bpasero this is exactly as @weeblr wrote. In vscode UserVoice there are many requests for auto-saving untitled files as it's available in another editors (Notepad++, SublimeText, Atom, etc...) - _without asking for a filename_, for example:

- [Auto save temporary files](https://visualstudio.uservoice.com/forums/293070-visual-studio-code/suggestions/7756146-auto-save-temporary-files)

Is this feature in consideration to implement in next releases?

**Comment by victorhooi** [comment]
Yup, this would be super awesome for me as well! =).

I often use my text-editors as a scratch-pad.

Sublime Text has had this for a while, and Atom had it via third-party extensions (e.g. [save-session](https://atom.io/packages/save-session). However, Atom recently rolled it into the core - perhaps some of their discussion might be helpful to our implementation?

Relevant PR - https://github.com/atom/atom/pull/9968

**Comment by LeviDettwyler** [comment]
Incredibly disappointed that this wasn't included in the initial release. Mac users are used to essentially all of their applications working this way. I just lost a lot of work because I incorrectly assumed that Microsoft was making this app Mac-friendly. This is not beta software, this is production software. Users are going to lose data because this is a **standard feature** for Mac document / code editors, Sublime included. There are very few excuses for an application to lose unsaved user data in the event of an unexpected application termination.

**Comment by weeblr-dev** [comment]
@bpasero I just read issue #101 linked above by @katopz, and I think I understand why you would not favor this change. This maybe a misunderstanding.

>  I find it very confusing that you quit an editor and don't get prompted to save dirty files. You might expect all changes are written to disk when in fact they are not.

No, we don't expect changes changes to be written to files. We expect changes, ie the current editor content, to be saved. Nobody said it was to be saved in the actual file being edited. On top of possibly failing, as you hint at, this might not be what you want (ie you made a change, but this was no good and you actually don't want to keep it)

So the point is save the "state" of the editor, in a separate area probably (and please, save that data in a portable way, not in the user folder #329 !).

> In the VS Code team we all run with auto save enabled (File | Auto Save) so we are past these issues it seems :)

You're probably actually behind what others do. Saving on the fly is common feature in many IDE and editors, but basically, it's still missing what this feature request is about. To auto-save, you need to have a file name. There are many situations where we don't want to have to choose a folder, and make up a name, and then have that useless file lying around.
Auto Save is fine. I don't use it personally, I rather select what needs to be saved and when, but I hate being prompted for file names, when I don't have to.

Rgds

**Comment by bpasero** [comment]
@weeblr-dev I have removed my comments on https://github.com/Microsoft/vscode/issues/101 because they do not reflect our latest thinkings. Sorry for the confusion. It is a perfectly valid feature request.

I would still close this one as duplicate of https://github.com/Microsoft/vscode/issues/101 though if you are ok with it?

**Comment by weeblr-dev** [comment]
@bpasero Yes, sure. I thought about that when I realized #101 was created some times before this one.

**Comment by bpasero** [comment]
Ok good.

**Comment by lovebdsobuj** [comment]
This feature `"Unsaved files should be preserved on quit"` still not yet implement in **VSCode** version 1.4 !
It's really needed feature.

**Comment by erhhung** [comment]
I was shocked that VSCode did not support autosaving files when I just lost a lot of work after a crash and was expecting my files to be automatically recovered. I looked around and found that this **feature is available, but is not enabled by default!**

What bad UX for a user to discover this preference only after losing work! Just about every other code editor I use (Emacs included), have autosave enabled by default.

Now, the autosave feature in some editors writes to separate files, so the user still has the option of not recovering. VSCode can implement that behavior as another preference setting.

**Comment by erhhung** [comment]
Having now explicitly set my `files.autoSave` preference to `afterDelay`, I realized that this **in-place save** really doesn't work in common web-development situations where you run a **watcher that rebuilds code on-the-fly**.

None of the options for that preference works in that scenario--only the tried-and-true explicit commit (save or save all) coupled with **autosave to temp files** will work, IMO.

BTW, my recommendation on the autosave behavior would not be a duplicate of #101 as it's more general and has nothing to do with quitting the editor.
